don't escape slash in toJSON (issue 17587)

aG0aep6G opened the pull request:
On input, both escaped and unescaped slashes must be recognized. They don't
need to be escaped on output.

Fixes issue 17587 (enhancement request).

----

As far as I see, the argument for escaping slashes is that it turns `"</script>"` into `"<\/script>"`. The latter can be put into JavaScript code without breaking it. But one can still emit stuff like `"<!-- <script>"` which confuses JS engines. So this seems a weak argument.

adamdruppe commented:
> But one can still emit stuff like "<!-- <script>" which confuses JS engines. So this seems a weak argument.

That just breaks the script though; the HTML parser sees it as all commented out until the next >, which is still inside the Javascript string.

But `</script>` allows you to escape the context altogether.

(using php for the sample just cuz it demos the idea)

```
<!DOCTYPE html>
<html>
<head>
<script>
var result = <?php
      // php's json_encode does the \/ thing too, so using str_replace to undo it to simulate the change in this PR
        echo str_replace("\\/", "/",
                json_encode("</script><script>alert('pwned');</script>"));
?>;
alert(result);
</script>
</head><body></body></html>
```

The user string is executed as code... pwned.

You can't do something like that with `<!--`. You, as the user content author, can make the parser think it is all comments and break the script (and yeah the website author ought to consider that case and do something about it), but you can't actually get it to run the XSS payload (well, at least not that I know of).

It is the difference between a merely broken page when the coder just lazily calls the generic function and a security-compromised page when he does the same. Who knows if there's D code out there which uses toJSON in this sense today.... do you want to give them a silent security regression?

I don't mind having an option to suppress the behavior, but let's not change the default to "silently enable XSS attacks".

aG0aep6G commented:
> You can't do something like that with <!--. You, as the user content author, can make the parser think it is all comments and break the script (and yeah the website author ought to consider that case and do something about it), but you can't actually get it to run the XSS payload (well, at least not that I know of).

Good point. Reverted the default back to escaping. Added `JSONOptions.doNotEscapeSlashes` (not sure about the name). If this is the way to go, I'd squash the commits.

aG0aep6G commented:
Rebased. Squashed the commits.

CyberShadow commented on std/json.d:
```diff
@@ -1153,7 +1154,13 @@ string toJSON(const ref JSONValue root, in bool pretty = false, in JSONOptions o
+                if (!(options & JSONOptions.doNotEscapeSlashes))
+                    json.put('\\');
+                json.put('/');
+                break;
```
Indentation looks off

CyberShadow commented on std/json.d:
```diff
@@ -75,7 +75,8 @@ enum JSONOptions
     specialFloatLiterals = 0x1, /// encode NaN and Inf float values as strings
-    escapeNonAsciiChars = 0x2   /// encode non ascii characters with an unicode escape sequence
+    escapeNonAsciiChars = 0x2,   /// encode non ascii characters with an unicode escape sequence
+    doNotEscapeSlashes = 0x4, /// do not escape slashes ('/')
```
Negative options are ungood, but I can't immediately think of a better name. Maybe `rawSlashes`?

Comments got out of alignment BTW

aG0aep6G commented:
> Negative options are ungood,

Agreed. Makes the code not unhard to read, too.

> but I can't immediately think of a better name.

Same here.

> Maybe rawSlashes?

Doesn't strike me as an improvement. Gets rid of the negative, but isn't as clear. I wouldn't oppose it, though.

> Comments got out of alignment BTW

> Indentation looks off

Fixed.

CyberShadow commented:
> Doesn't strike me as an improvement. Gets rid of the negative, but isn't as clear. I wouldn't oppose it, though.

PHP has `JSON_UNESCAPED_SLASHES`. How about `unescapedSlashes`, then?

aG0aep6G commented:
> PHP has JSON_UNESCAPED_SLASHES. How about unescapedSlashes, then?

PHP also has `JSON_UNESCAPED_UNICODE` where we have `escapeNonAsciiChars`. `doNotEscapeSlashes` at least keeps it active.

I've also considered `noEscapeSlashes` (ungrammatical, I think) and `dontEscapeSlashes` (maybe too sloppy).

aG0aep6G commented:
Ping. @CyberShadow, I don't care strongly about the name. `doNotEscapeSlashes` seems the best to me, but if you'd prefer another one, just say so and I'll change it.

CyberShadow commented:
I like `unescapedSlashes` more despite the inconsistency, but I also don't really care enough to ask you to change it.

CyberShadow commented:
Are things like enum members subject to Andre's new symbol approval review?
